### PR TITLE
upstream: allowing splitting out choosing host from choosing connecti…

### PR DESCRIPTION
**Pull request by alyssawilk**
This is the first step in a planned refactor to allow asynchronous host selection

Risk Level: high (core code refactor)
Testing: existing tests pass
Docs Changes: n/a
Release Notes: n/a (until async selection is supported)

https://github.com/envoyproxy/envoy/issues/36819

### CODE REVIEW

**Comment by alyssawilk** [comment]
We do now :-)
https://github.com/envoyproxy/envoy/issues/36819

**Comment by alyssawilk** [comment]
feel free to wait on this until Monday to give folks a time to comment on the issue (since as you say it's a large refactor)

**Comment by wbpcode** [comment]
I foget this PR last week, will try to review this week. Is this work still have priority? cc @alyssawilk

**Comment by alyssawilk** [comment]
I'm still game to implement it, yes

**Review by wbpcode — COMMENTED** [review]
LGTM with these changes. Not sure will this bring some performance penalty, but it would be different topic.

But considering this is a huge change to our core code base, it would be better to let another senior maintainer to give a second pass. Thanks.

And please resolve the confliction. :)

/wait

File: source/common/router/router.cc
Comment (RyanTheOptimist) [inline source/common/router/router.cc:868]: nit: auto => HostConstSharedPtr to be explicit.

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (RyanTheOptimist) [inline source/common/tcp_proxy/tcp_proxy.cc:572]: ditto

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (RyanTheOptimist) [inline source/common/tcp_proxy/tcp_proxy.cc:577]: In the case where hose is null, should we reset `generic_conn_pool_`? I'm looking at the old code and it seems to unconditionally assign to  `generic_conn_pool_`

**Review by RyanTheOptimist — COMMENTED** [review]
LGTM, modulo a few small comments

File: source/common/tcp_proxy/tcp_proxy.cc
Comment (alyssawilk) [reply source/common/tcp_proxy/tcp_proxy.cc:577]: yeah I was torn on this.  on the one hand I think it doesn't make sense to try to create a generic connection pool wrapping a known nullptr host as it's destined to fail.  on the other hand it would make it a slightly lower risk refactor.  have a preference?

Comment (RyanTheOptimist) [reply source/common/tcp_proxy/tcp_proxy.cc:577]: Not a strong one, but since this isn't runtime guarded, I would probably be inclined to err on the side of lower risk. But I do agree that it's likely to fail either way.

Comment (alyssawilk) [reply source/common/tcp_proxy/tcp_proxy.cc:577]: Ah so  I took another look and what would happen and what would happen is the conn pool created would fail validity checks in createGenericConnPool and we'd return nullptr, so this is equivalen and avoids some work

**Review by RyanTheOptimist — APPROVED** [review]
LGTM!

**Pull request merged**
